### ISSUE TITLE
Feature Request: Generic Automodel CP interface for downstream libraries e.g. NeMo RL

### ISSUE THREAD

**Issue by sharonyu-115** [NVIDIA-NeMo/Automodel#2861]
## Is Your Feature Request Related To A Problem? Please Describe.

Downstream training libraries such as NeMo RL need to use Automodel context parallelism (CP), but  <br>today they must understand and partially reimplement Automodel's CP internals. The current APIs expose several related but different paths:

* Generic SDPA CP via `create_context_parallel_ctx`.
* Transformer Engine CP, including THD or packed batch layouts.
* Model-owned CP via `prepare_model_inputs_for_cp` and `_cp_make_batch_fn` such as for gemma4.

For example, NeMo-RL can use the generic `create_context_parallel_ctx` path for standard SDPA-style CP. For models such as Gemma4, however, NeMo-RL needs a separate model-owned CP path that calls `prepare_model_inputs_for_cp`, invokes `make_cp_batch_and_ctx`, infers the sequence shard layout, computes local next-token targets, and implements its own token-level gather.

This is brittle for downstream libraries because `prepare_model_inputs_for_cp` is not a reliable capability signal by itself. Some models expose that method only to precompute full-sequence multimodal embeddings before generic CP sharding. Other models use it to attach a model-owned CP sharder and attention implementation. Downstream libraries should not need to infer those semantics from method names or private batch keys.

It would be good to have a public Automodel-owned CP contract that tells downstream libraries:

* Which CP backend was selected.
* Which sequence layout is being used.
* Which local sequence range each rank owns.
* How to gather token-level or sequence-level outputs correctly.
* Which context manager and model inputs should be used for the forward.

## Describe The Solution You'd Like

Add a public Automodel CP preparation interface for downstream training libraries. The interface should accept a model, device mesh, batch, etc., then return a stable CP plan object. So the general idea is that Automodel exposes the CP layout and helper operations needed for NeMo RL to compute losses and logprobs without branching on private CP implementation details.

For example:

```python
cp_plan = prepare_cp_forward(
    model=model,
    device_mesh=device_mesh,
    batch=batch,
    purpose="causal_lm_forward",
)

with cp_plan.forward_context():
    outputs = model(**cp_plan.model_inputs)

local_logits = extract_logits(outputs)
local_targets = cp_plan.local_next_token_targets(
    input_ids=batch["input_ids"],
    seq_dim=1,
)

local_token_logprobs = compute_local_token_logprobs(local_logits, local_targets)
token_logprobs = cp_plan.gather_token_tensor(local_token_logprobs, seq_dim=1)
```

The returned plan may possibly expose:

```python
class CPForwardPlan:
    backend: Literal["none", "sdpa", "te", "model_owned"]
    layout: Literal["none", "torch_load_balanced", "contiguous_bsh", "packed_thd", "custom"]

    model_inputs: dict[str, Any]
    original_seq_len: int
    padded_seq_len: int
    local_seq_len: int
    local_seq_start: int | None
    local_seq_end: int | None
    seq_dim: int
    cp_rank: int
    cp_size: int

    def forward_context(self) -> ContextManager[Any]: ...

    def local_next_token_targets(
        self,
        input_ids: torch.Tensor,
        seq_dim: int = 1,
        pad_token_id: int = 0,
    ) -> torch.Tensor: ...

    def gather_token_tensor(
        self,
        local_tensor: torch.Tensor,
        seq_dim: int = 1,
        *,
        trim_to_original: bool = True,
        differentiable: bool = True,
    ) -> torch.Tensor: ...

    def gather_sequence_tensor(
        self,
        local_tensor: torch.Tensor,
        seq_dim: int = 1,
        *,
        trim_to_original: bool = True,
        differentiable: bool = True,
    ) -> torch.Tensor: ...
```

The exact class shape can differ, but the contract should make these semantics  <br>explicit:

* Whether CP is disabled, generic SDPA, TE, or model-owned.
* Whether the sequence layout is contiguous, torch load-balanced, packed THD, or  <br>custom.
* How local sequence positions map back to global sequence positions.
* Whether gather helpers are differentiable.
* Whether gather helpers return rank-ordered/global sequence order.
* Whether helpers can gather cheap token-level tensors such as `[B, S/cp]`  <br>without materializing full logits `[B, S, V]`.

The token-level gather requirement is important for RL libraries. For long  <br>contexts, gathering full logits can erase CP's memory benefit. NeMo-RL needs to  <br>compute local logprobs on `[B, S/cp, V]` and gather only token-level tensors  <br>such as `[B, S/cp]`.

## Describe Alternatives You've Considered.

One alternative is for downstream libraries to keep manually branching:

* Generic SDPA CP via `create_context_parallel_ctx`.
* TE CP via TE-specific setup and `make_cp_batch_for_te`.
* Model-owned CP via `prepare_model_inputs_for_cp`, `_cp_make_batch_fn`, and  <br>model-specific layout assumptions.

This works, but it duplicates Automodel knowledge outside Automodel and is easy to break as models add custom CP behavior. It also forces downstream libraries to infer model-owned CP support from implementation details such as method presence or private batch keys.

## Additional Context.

This came up while integrating Gemma4 Automodel CP into NeMo-RL, for which I had to add a special model-owned CP path on top of the current generic SPDA path [https://github.com/sharonyu-115/RL/tree/gemma4-cp](<https://github.com/sharonyu-115/RL/tree/gemma4-cp>), but no unified way to consume Automodel's TE CP or future model-specific CP variants.

In the Gemma4 integration, NeMo-RL currently has to:

* Detect model-owned CP.
* Disable its generic CP iterator/context path.
* Call `prepare_model_inputs_for_cp`.
* Let `make_cp_batch_and_ctx` dispatch to `_cp_make_batch_fn`.
* Assume Gemma4's rank-ordered contiguous sequence layout.
* Compute local next-token targets from full `input_ids`.
* Gather token-level logprobs with a custom contiguous all-gather.

Those are exactly the pieces that would be cleaner if Automodel returned an explicit `CPForwardPlan`.

Related upstream work:

* PR NVIDIA-NeMo/Automodel#1914: Gemma4 model-owned CP: [https://github.com/NVIDIA-NeMo/Automodel/pull/1914](<https://github.com/NVIDIA-NeMo/Automodel/pull/1914>)
* Issue NVIDIA-NeMo/Automodel#2340: Automodel roadmap, especially "Provide an RL Library Interface" and "Context Parallelism and Packing": [https://github.com/NVIDIA-NeMo/Automodel/issues/2340](<https://github.com/NVIDIA-NeMo/Automodel/issues/2340>)
* NeMo RL Gemma4 CP draft: [https://github.com/sharonyu-115/RL/tree/gemma4-cp](<https://github.com/sharonyu-115/RL/tree/gemma4-cp>)

**Comment by sharonyu-115** [comment]
Hi @zpqiu as you worked on the dsv4 cp support integration to nemo rl, would you review if this high-level idea makes sense to you from nemo rl perspective? any comments? Thanks!

**Comment by zpqiu** [comment]
Overall +1 from my side. The DSV4 CP integration hit exactly these pain points — we ended up with a separate model-specific path (own batch sharding, target alignment, token-level gather) next to the generic one, duplicating layout knowledge downstream. The `CPForwardPlan` split feels right: forward and loss stay in the training library, layout semantics stay in Automodel.

One suggestion: besides gather, it would help if the plan also covered the shard/scatter direction. RL losses consume per-token tensors that aren't model inputs (advantages, old/ref logprobs, loss masks), and they must be sharded identically to input_ids since we compute the loss on local shards and only reduce across ranks. Could be a plan.shard_token_tensor(tensor, seq_dim=1), or simply a contract that token-aligned entries in batch get co-sharded — exact API shape up to you all, just flagging it. Thanks!

**Comment by akoumpa** [comment]
Hi @sharonyu-115  and @zpqiu ,

Thank you for making this issue and sharing the API interface!

I wanted to ask you if the `backend` and `layout` attributes in `CPForwardPlan` should be user-visible? I don't have 100% visibility of how it's intended to be used in RL, so I'm wondering if these are necessary or we could rely on `CPForwardPlan`'s methods instead?

My reasoning is that those could then introduce additional value-based dispatching on the RL side and perhaps we want to avoid it, but please let me know if that's case since you have the full visibility of the use intention.

Thank you. 

```
class CPForwardPlan:
    backend: Literal["none", "sdpa", "te", "model_owned"]
    layout: Literal["none", "torch_load_balanced", "contiguous_bsh", "packed_thd", "custom"]
```

**Comment by sharonyu-115** [comment]
> Hi [@sharonyu-115](https://github.com/sharonyu-115) and [@zpqiu](https://github.com/zpqiu) ,
> 
> Thank you for making this issue and sharing the API interface!
> 
> I wanted to ask you if the `backend` and `layout` attributes in `CPForwardPlan` should be user-visible? I don't have 100% visibility of how it's intended to be used in RL, so I'm wondering if these are necessary or we could rely on `CPForwardPlan`'s methods instead?
> 
> My reasoning is that those could then introduce additional value-based dispatching on the RL side and perhaps we want to avoid it, but please let me know if that's case since you have the full visibility of the use intention.
> 
> Thank you.
> 
> ```
> class CPForwardPlan:
>     backend: Literal["none", "sdpa", "te", "model_owned"]
>     layout: Literal["none", "torch_load_balanced", "contiguous_bsh", "packed_thd", "custom"]
> ```

Hi @akoumpa, Thanks for your review and comments. I think the `backend` should be automatically determined by Automodel and made transparent to RL for simplicity. While the sharding layout must be exposed as RL needs to compute the logprob with the local logits on each rank. We've reviewed the implementation from @HuiyingLi, the interface looks good. We don't necessarily need to have the plan layer, as long as the shader is returned we should be good.

**Comment by akoumpa** [comment]
Thank you @sharonyu-115 , i realized after the fact that @HuiyingLi will go with another interface, my goal is to ensure we have the least amount of fields / functions in this public API in order to ensure maintainability and/or make sure we don't (accidentally) add something that's unnecessary, I am not saying the current API does this, just trying to understand the current design.

@HuiyingLi when you get the chance, please feel free to share the API you settled on.

Thank you all 🙇

**Comment by HuiyingLi** [comment]
Strongly agree with this — a single CP contract is the right call, and it's what #2937 builds. A couple of thoughts to sharpen the shape, split into (1) what the shared object should be, and (2) how downstream libraries should actually consume it.

## 1. What the CP object is

Under the hood, every CP path here reduces to one fact: **which global sequence position does each local token map to.** Round-robin, TE/THD, and model-owned CP differ only in *how that mapping is computed* — but once you have it, everything else is derived from it:

- **shard (down):** a full-sequence tensor → this rank's local slice, in the same layout the model inputs were split into;
- **gather (up):** a local slice → the full sequence again (all-gather + reorder).

So the contract really only needs to expose that one mapping plus those two operations. That directly gives you what this issue asks for — `backend`/`layout` become descriptive fields (no more guessing from method names or private batch keys), and `gather_token_tensor` works on `[B, S/cp]` **without ever materializing `[B, S, V]`**, which is the whole point for long-context RL.

One important addition: make it **symmetric** — RL doesn't only *gather* logprobs up, it also needs to *shard its own* token-aligned tensors down (advantages, old-logprobs, loss mask) so they line up with the local logits before the loss. So the contract should offer both `shard_token_tensor` (down) and `gather_token_tensor` (up), both differentiable. (#2937 does exactly this.)

## 2. How downstream should consume it

There are really **two consumer shapes**, and the same object serves both:

**A. The caller runs its own forward loop** (this issue's example — e.g. a policy worker that calls `model(**cp_plan.model_inputs)` itself). Here the plan is **returned to the caller**, and the caller invokes `shard_token_tensor` / `gather_token_tensor` around its own forward. This is the interface the issue describes, and it's the right one for a worker that owns its forward.

**B. The caller goes through Automodel's training entry** (`Engine.forward_backward` / the recipe step). Here the caller should **not have to touch the plan at all.** The Engine resolves the plan internally, shards the caller's *declared* token tensors down, runs forward + loss on the shard, and gathers logprobs back up before returning — so the caller gets full-sequence outputs and never sees CP. This keeps the algorithm layer free of any CP/topology awareness.

Both are the **same plan object** — the only difference is *who* calls shard/gather: the downstream library (A) or the Engine (B). So I'd suggest the contract be designed so it's equally callable by an external forward loop and by Automodel's own step, rather than assuming only the external case. That way #2861 is satisfied for callers that own their forward, and callers that go through Automodel get CP for free without reimplementing anything.

**Comment by HuiyingLi** [comment]
The contract from this issue shipped in #2937 (validated end-to-end — [parity ledger](https://github.com/NVIDIA-NeMo/Automodel/pull/2937#issuecomment-5022113503)). Code samples for the NeMo-RL integration, mapped to the asks in the issue body:

**1. One forward path, no backend branching** — replaces `create_context_parallel_ctx` / TE setup / `prepare_model_inputs_for_cp` + `_cp_make_batch_fn` detection:

```python
from nemo_automodel.components.distributed.cp_utils import prepare_cp_forward

ctx, batch, sharder = prepare_cp_forward(
    model, device_mesh, batch,
    padding_token_id=tokenizer.pad_token_id,
)
cp_mesh = device_mesh["cp"]

with ctx():
    out = model(**batch)          # batch arrives already sharded to [B, S/cp]
```

The same three lines cover generic SDPA CP, TE/THD, MagiAttention, and model-owned CP (gemma4, dsv4, glm) — resolution order is model-owned > magi > TE > generic. At `cp_size == 1` you get an identity sharder, so the consumer code is branch-free across CP on/off.

**2. Local next-token targets / co-sharded token streams** — instead of a dedicated `local_next_token_targets`, shard any full-length token-aligned tensor *exactly like the model inputs* (`fill` pads the CP-padding tail):

```python
full_targets = torch.roll(input_ids, shifts=-1, dims=1)          # caller coordinates, original S
local_targets    = sharder.shard_token_tensor(cp_mesh, full_targets, fill=pad_id)
local_advantages = sharder.shard_token_tensor(cp_mesh, advantages,  fill=0.0)
```

Alignment is by construction: the verb replays the same pad / flatten (THD) / reposition (DSV4 packed repad) transform the batch went through, and a length that cannot be transformed unambiguously **raises** instead of silently misaligning.

**3. Token-level gather without materializing `[B, S, V]`** (the RL memory ask):

```python
local_logprobs = torch.gather(
    local_logits.log_softmax(-1), -1, local_targets.unsqueeze(-1)
).squeeze(-1)                                                     # [B, S/cp]

token_logprobs = sharder.gather_token_tensor(cp_mesh, local_logprobs, trim=True)
# [B, S_original] — global sequence order, CP padding trimmed
```

The gather is differentiable (`torch.distributed.nn.functional.all_gather`, gradients route back to the owning ranks); its collective semantics, including the sum-over-consumers gradient, are pinned by `tests/functional_tests/context_parallel/run_cp_sharder_token_verbs.py`.

**4. Layout introspection** — instead of `backend` / `layout` enums (dropped per the discussion above so nothing can branch on them), the layout is exposed behaviorally:

```python
idx = sharder.local_token_global_indices(cp_mesh, padded_seq_len, device)
# LongTensor: global position of every local token (closed-form for
# contiguous / round-robin; data-dependent layouts install theirs at shard time)

layout = sharder.shard_layout    # original_seq_len, padded_seq_len,
                                 # input_row_shape (THD), repad position map (DSV4)
```

Checklist from the issue body: which context manager + model inputs → returned by the call; which local range each rank owns → `local_token_global_indices`; how to gather correctly → `gather_token_tensor(trim=True)`; which backend/layout → deliberately not a tag, the verbs make it irrelevant. Happy to review the NeMo-RL side migration off the `gemma4-cp` branch.

**Comment by akoumpa** [comment]
Thanks @HuiyingLi for the implementation and for spelling out the symmetric shard/gather requirement in #2937.

I took another pass at the API and pushed a focused follow-up here: https://github.com/NVIDIA-NeMo/Automodel/tree/akoumpa/refactor/cp-sharder-ownership

The proposed call shape is now:


**Before**

```python
from nemo_automodel.components.distributed.cp_utils import make_cp_batch_and_ctx

cp_size = getattr(self.dist_setup, "cp_size", self.cfg.get("distributed.cp_size", 1))
model = self.model_parts[0]

if self.magi.enabled:
    train_ctx, batch = self.magi.prepare_llm_batch(
        model,
        batch,
        device_mesh=self.device_mesh,
        is_thd=_thd_collater,
        pad_id=self.tokenizer.pad_token_id if self.tokenizer else 0,
        num_chunks=num_chunks,
    )
else:
    if cp_size > 1 and hasattr(model, "prepare_model_inputs_for_cp"):
        # Old model-owned path pushed private CP details into the batch.
        batch.update(
            model.prepare_model_inputs_for_cp(
                input_ids=batch["input_ids"],
                num_chunks=num_chunks,
            )
        )

    train_ctx, batch = make_cp_batch_and_ctx(
        self.device_mesh,
        batch,
        use_te=_thd_collater,
        padding_token_id=self.tokenizer.pad_token_id if self.tokenizer else 0,
        num_chunks=num_chunks,
    )

with train_ctx():
    outputs = model(**batch)
```

**After**

```python
from nemo_automodel.components.distributed.context_parallel import ContextParallelSharder

cp_sharder = ContextParallelSharder(
    model,
    device_mesh,
    batch,
    padding_token_id=tokenizer.pad_token_id,
)
train_ctx, batch = cp_sharder.shard(batch)

local_advantages = cp_sharder.shard_token_tensor(
    advantages,
    fill=0.0,
)

with train_ctx():
    outputs = model(**batch)
    local_logprobs = compute_token_logprobs(outputs, batch["labels"])

full_logprobs = cp_sharder.gather_token_tensor(
    local_logprobs,
    trim=True,
)
```

The CP implementation is grouped under `nemo_automodel/components/distributed/context_parallel/`, with `ContextParallelSharder` as the package-level API. This removes `prepare_cp_forward` and renames `ContextParallelismSharder` to `ContextParallelSharder`. The constructor resolves the attention backend from the live model, reads the packed layout from `batch["qkv_format"]`, invokes model-owned CP preparation when available, and binds the CP/TP meshes. `shard(batch)` performs the actual batch sharding, records the resulting token layout, and returns `(train_ctx, sharded_batch)`. There is intentionally no separate `.context()` call; the returned context factory is used directly with `with train_ctx():`.

The token helpers no longer take `device_mesh["cp"]`; they use the mesh and layout already bound to the sharder. Additional token-aligned tensors therefore use `shard_token_tensor(...)` and `gather_token_tensor(...)` directly instead of being inserted into the batch.

I would keep this as a small layout-aware wrapper rather than returning a `DTensor`: standard `Shard(dim)` placements cannot represent torch CP head/tail ordering, packed THD partitions, Magi dispatch permutations, or model-owned re-padding.
